### PR TITLE
feat(tracking) Maksimchepelev/feature tracking 428

### PR DESCRIPTION
**Pull request by MaksimChepelev**

### CODE REVIEW

File: daemon/src/org/eclipse/iofog/field_agent/FieldAgent.java [hunk 1]
```diff
@@ -301,6 +312,9 @@ private void verificationFailed(Exception e) {
                             processRoutes(microservices);
                             MessageBus.getInstance().update();
                         }
+
+                        Tracker.getInstance().handleEvent(TrackingEventType.MICROSERVICE,
+                                loadMicroservicesJsonFile());
```
Comment (MaksimChepelev) [inline daemon/src/org/eclipse/iofog/field_agent/FieldAgent.java]: try not to read file

File: daemon/src/org/eclipse/iofog/field_agent/FieldAgent.java [hunk 2]
```diff
@@ -223,6 +226,14 @@ private boolean notProvisioned() {
         }
     };
 
+    public final void postTracking(JsonObject events) {
+        try {
+            orchestrator.request("tracking", RequestType.POST, null, events);
+        } catch (Exception e) {
+            logError("Unable send tracking logs : " + e.getMessage(), e);
```
Comment (Railag) [inline daemon/src/org/eclipse/iofog/field_agent/FieldAgent.java]: unable to

File: daemon/src/org/eclipse/iofog/field_agent/FieldAgent.java [hunk 3]
```diff
@@ -473,6 +487,12 @@ private void processRoutes(List<Microservice> microservices) {
         microserviceManager.setRoutes(routes);
     }
 
+    private JsonArray loadMicroservicesJsonFile() {
+        String filename = "microservices.json";
```
Comment (Railag) [inline daemon/src/org/eclipse/iofog/field_agent/FieldAgent.java]: const

Comment (MaksimChepelev) [inline daemon/src/org/eclipse/iofog/field_agent/FieldAgent.java]: constant

File: daemon/src/org/eclipse/iofog/tracking/Tracker.java [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        if (instance == null) {
+            synchronized (Tracker.class) {
+                if (instance == null)
+                    instance = new Tracker();
+            }
+        }
+        return instance;
+    }
+
+    private String id;
```
Comment (Railag) [inline daemon/src/org/eclipse/iofog/tracking/Tracker.java]: id -> uuid

File: daemon/src/org/eclipse/iofog/tracking/Tracker.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    Timer loggerTimer = null;
```
Comment (MaksimChepelev) [inline daemon/src/org/eclipse/iofog/tracking/Tracker.java]: private?

File: daemon/src/org/eclipse/iofog/tracking/Tracker.java [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        public void run() {
+            List<TrackingEvent> events = TrackingEventsStorage.getInstance().popAllEvents();
+            JsonArrayBuilder jsonArrayBuilder = Json.createArrayBuilder();
+
+            events
+                    .forEach(el -> {
+                        jsonArrayBuilder.add(el.toJsonObject());
+                    });
+
+            //TODO not tested with server
```
Comment (Railag) [inline daemon/src/org/eclipse/iofog/tracking/Tracker.java]: clear

File: daemon/src/org/eclipse/iofog/tracking/TrackingEventType.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package org.eclipse.iofog.tracking;
+
+public enum TrackingEventType {
+    START("application started"),
+    PROVISION("provision"),
+    DEPROVISION("deprovision"),
+    CONFIG("config updated"),
+    TIME("running time"),
+    MICROSERVICE("microservices was updated"),
```
Comment (Railag) [inline daemon/src/org/eclipse/iofog/tracking/TrackingEventType.java]: were

File: daemon/src/org/eclipse/iofog/tracking/Tracker.java [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+        return randString.toString();
+    }
+
+    private List<String> getAllMacs() throws SocketException {
+        List<String> macs = new ArrayList<>();
+        Enumeration<NetworkInterface> networks = NetworkInterface.getNetworkInterfaces();
+        NetworkInterface inter;
+        while (networks.hasMoreElements()) {
+            inter = networks.nextElement();
+            if (inter.isVirtual()) {
```
Comment (MaksimChepelev) [inline daemon/src/org/eclipse/iofog/tracking/Tracker.java]: if !virtual

File: daemon/src/org/eclipse/iofog/tracking/TrackingInfoUtils.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+import javax.json.JsonObject;
+
+public class TrackingInfoUtils {
+    public static JsonObject getStartTrackingInfo() {
+        String gpsCoordinates = Configuration.getGpsCoordinates();
+        String gpsMode = Configuration.getGpsMode().name();
+        boolean developerMode = Configuration.isDeveloperMode();
+        String networkInterface = Configuration.getNetworkInterfaceInfo();
+        String version = CmdProperties.getVersion();
+        String agentStatus = StatusReporter.getFieldAgentStatus().getControllerStatus().name().toLowerCase();
```
Comment (Railag) [inline daemon/src/org/eclipse/iofog/tracking/TrackingInfoUtils.java]: wrap in try/catch

File: daemon/src/org/eclipse/iofog/tracking/Tracker.java (hunk 6 above)
Comment (MaksimChepelev) [inline daemon/src/org/eclipse/iofog/tracking/Tracker.java]: delete todo

File: daemon/src/org/eclipse/iofog/tracking/Tracker.java [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+                //send directly
+                //TODO add event that says about connection problem on controller
+                StringEntity requestEntity = new StringEntity(eventsListObject.toString(), ContentType.APPLICATION_JSON);
+                postMethod = new HttpPost("https://analytics.iofog.org/post");
+                postMethod.setEntity(requestEntity);
+
+                try {
+                    HttpResponse response = httpClient.execute(postMethod);
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Comment (MaksimChepelev) [inline daemon/src/org/eclipse/iofog/tracking/Tracker.java]: logger

File: daemon/src/org/eclipse/iofog/tracking/Tracker.java [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+                    .add("events", jsonArrayBuilder.build())
+                    .build();
+
+            HttpPost postMethod = null;
+            if (StatusReporter.getFieldAgentStatus().getControllerStatus().equals(Constants.ControllerStatus.OK)) {
+                //send to controller
+                FieldAgent.getInstance().postTracking(eventsListObject);
+            } else {
+                //send directly
+                //TODO add event that says about connection problem on controller
```
Comment (MaksimChepelev) [inline daemon/src/org/eclipse/iofog/tracking/Tracker.java]: delete

File: daemon/src/org/eclipse/iofog/tracking/TrackingEventsStorage.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    private List<TrackingEvent> events = new ArrayList<>();
+
+    protected synchronized void pushEvent(TrackingEvent event) {
+        events.add(event);
+    }
+
+    protected synchronized List<TrackingEvent> popAllEvents() {
+        List<TrackingEvent> subList = events.subList(0, events.size());
```
Comment (MaksimChepelev) [inline daemon/src/org/eclipse/iofog/tracking/TrackingEventsStorage.java]: simplify

File: daemon/src/org/eclipse/iofog/tracking/TrackingEventsStorage.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+public class TrackingEventsStorage {
+    private static TrackingEventsStorage instance = null;
+    public static TrackingEventsStorage getInstance() {
+        if (instance == null) {
+            synchronized (TrackingEventsStorage.class) {
+                if (instance == null)
+                    instance = new TrackingEventsStorage();
+            }
+        }
+        return instance;
```
Comment (epankou) [inline daemon/src/org/eclipse/iofog/tracking/TrackingEventsStorage.java]: You use TrackingEventsStorage only inside Tracker. I think you don't need singleton here. Just create reference to it in Tracker.

**Pull request merged**